### PR TITLE
Improve Windows VM examples

### PR DESCRIPTION
**Pull request by fabiand**
This change

- Changes the template to leverage presets
- Adds a VM only (no template) example which can be used in K8s

```release-note
Add a Microsoft Windows example for stock Kubernetes usage
```

### CODE REVIEW

**Comment by fabiand** [comment]
Orthorgonal to this is that we should test the examples.

**Comment by cynepco3hahue** [comment]
@fabiand you need to edit only `vms-generator` and run `make generate`

**Comment by fabiand** [comment]
_only_ is a tiny word in this case for the amount of work to do :)

**Comment by fabiand** [comment]
@urabenst could confirm that these manifests work to launch a windows vm.

File: cluster/examples/vm-windows2012r2.yaml [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+      domain:
+        devices:
+          disks:
+          - disk:
+              bus: sata
+            name: pvcdisk
+            volumeName: pvcvolume
+          interfaces:
+          - bridge: {}
+            model: e1000
```
Comment (fabiand) [inline cluster/examples/vm-windows2012r2.yaml]: @booxter so, in testing we see that the model will not appear in the vmi when we start the vm.

Do you know why?

**Comment by urabenst** [comment]
As @fabiand said, I can launch Windows VMs using these manifests. I tried Windows Server 2016

File: cluster/examples/vm-windows2012r2.yaml (hunk 1 above)
Comment (booxter) [reply cluster/examples/vm-windows2012r2.yaml]: No I don't. Perhaps a bug or an old release you are using?

Comment (fabiand) [reply cluster/examples/vm-windows2012r2.yaml]: Thanks. Was my env. I setup 0.7.0 cleanly and everything works as expected.

See https://github.com/kubevirt/kubevirt/issues/1337

**Comment by fabiand** [comment]
retest this please

File: manifests/release/demo-content.yaml.in [hunk 2]
```diff
@@ -13,6 +13,7 @@ spec:
     resources:
       requests:
         memory: 2G
+    devices: {}
```
Comment (rmohr) [inline manifests/release/demo-content.yaml.in]: @fabiand I think that you needed that is also because of the wrong kubevirt version. Can you revert?

**Comment by fabiand** [comment]
Yep, I did revert.

**Comment by fabiand** [comment]
retest this please

**Comment by fabiand** [comment]
Why the heck do I need to run make generate?

I suppose a previous PR was missing this ...

**Comment by cynepco3hahue** [comment]
@fabiand We generate all our example from the code `vms-generator.go` and not via adding yamls, so you will need to add additional code to `vms-generator.go` and to run `make generate` to get new examples

**Comment by fabiand** [comment]
That's problematic IMHO.

Contirbutors need to write tons of code to provide nice examples.

What's your suggestion: Close this issue and request a coder to write code or consider a way to have plain text examples and allow users to contribute?

**Comment by cynepco3hahue** [comment]
@fabiand I remember we discussed it once already, so IIRC:
- we generate examples from the code to avoid the problem with outdated examples in case of API change

I understand that it can be a little bit harder to create it via code, but it still not so hard, also we always here to help :smile:

Also do you sure that windows2012 VM example is working I see it lack things like `clock` and `feature` sections?

**Comment by urabenst** [comment]
@cynepco3hahue Nothing much I could add about how you generate the yaml files, but the windows2012 example does definitely work for me

**Comment by rmohr** [comment]
@urabenst just to be sure, do you have our windows preset installed?

**Comment by urabenst** [comment]
@rmohr I am using this VirtualMachineInstancePreset and I just realized that it contains clock and feature sections
[preset.txt](https://github.com/kubevirt/kubevirt/files/2218390/preset.txt)

**Comment by rmohr** [comment]
That's what I thought. @fabiand should this be a standalone example, or an example which expects that people are using the preset?

**Comment by rmohr** [comment]
> That's what I thought. @fabiand should this be a standalone example, or an example which expects that people are using the preset?

Ok, to answer the question myself, from the PR description:

> Changes the template to leverage presets

**Comment by cynepco3hahue** [comment]
@rmohr @fabiand I do not see windows preset under example, so it must be part of the PR.

**Comment by fabiand** [comment]
I wonder if it was ever considered to do it in a more "natural" flow.

Thus: Have the plain text yamls, and feed them to kubevirt in order to validate them.

This would be much closer to the users flow. It would also strengthen the validation that a user is seeing.

**Comment by booxter** [comment]
@fabiand it's probably easier to share common pieces of config files in code between multiple templates. But if the decision is made to maintain examples directly in git then we would of course need to replace the current (arguably convoluted) way to validate specs with other means, like what you described. It's still a significant time investment to switch to a different approach, so not sure if it fits anyone's priorities.

**Comment by fabiand** [comment]
Right.
Do you recall what parts were intended to be shared between templates?

**Comment by booxter** [comment]
I wasn't there when `vms-generator` was written, but you can check the tool code for `add...` functions that manage disk sections, they seem to be examples of section reuse. BTW since we still have yaml files in git (and can regenerate them on demand anyway), we can enhance validation as you suggested without necessarily touching the tool, so these two proposals seem to be separate and should be judged that way.

**Comment by rmohr** [comment]
> I wonder if it was ever considered to do it in a more "natural" flow.
> 
> Thus: Have the plain text yamls, and feed them to kubevirt in order to validate them.
> 
> This would be much closer to the users flow. It would also strengthen the validation that a user is seeing.

@fabiand  Having it in code is in general much better fur us than as yaml:

* Easy to share config parts (we do that)
* Autocomplete for engineers when writing the examples (code completion)
* Functional tests can directly use the structs (we have tests for some of them, but @gszasz copied the structs instead of using them directly, that should be fixed)
* That the serialization goes wrong is very unlikely and we e.g. don't loose wrong written or unknown fields
* Finally the examples directory is not meant to contain dozens of examples. They all bring a maintenance burden, so it is not really a drawback that it is not that easy for arbitrary users to contribute examples here. I doubt that someone who has a hard time adding the examle is capable of writing and maintaining the tests for the example.

**Comment by fabiand** [comment]
I'm suggesting to keep presets and templates in a repository separate from KubeVirt.

I've started this repo https://github.com/fabiand/common-templates to gather presets and templates for common guests., including functional tests for templates and presets.

**Pull request closed without merging**